# DS n° 1

## Correction

#### Question 1: réponse d

Un octet permet de représenter exactement $2^8 = 256$ valeurs, échelonnées entre $0$ et $255$.

#### Question 2: réponse c

Dans le même ordre d'idée, on peut représenter $2^{16} = 65536$ valeurs sur 2 octets (16 bits), mais la plus grande de ces valeurs sera $65535$.

#### Question 3: réponse b

Demandons à python la réponse:

In [1]:
bin(153)

'0b10011001'

À la main, il était relativement aisé d'utiliser l'algorithme de conversion par les puissances de 2 décroissantes, et écrire
$$153 = 128 + 16 + 8 + 1$$

#### Question 4: réponse d

In [2]:
0b11010110

214

En effet, on trouve aisément (même sans calculatrice) à l'aide de regroupement astucieux:
\begin{align*}
128 + 64 + 16 + 4 + 2 & = (128 + 2) + (64 + 16) + 4  \\
& = 130 + 80 + 4 \\
& = 214
\end{align*}

#### Question 5: réponse c

In [3]:
0xad

173

Ici, le plus simple était d'écrire:
$$10\times 16 + 13 = 160 + 13 = 173$$

#### Question 6: réponse a (ou pas...)

Le programme proposé contient en fait un énorme bug (bien involontaire), que d'ailleurs personne (moi y compris) n'a remarqué: le résultat se trouvé dans la variable du même nom, pas dans la variable ```N``` qui est nécessairement nulle à la fin de la boucle ```while```, puisque celle-ci ne s'arrête qu'à cette condition.

In [4]:
N = 173
résultat = ""
while N > 0:
    if N % 2 == 0:
        résultat = "0" + résultat
        N = N // 2
    else:
        résultat = "1" + résultat
        N = (N - 1) // 2
N

0

In [5]:
résultat

'10101101'

In [6]:
bin(173)

'0b10101101'

On constate ici que ```résultat``` contient bien la représentation binaire de 173.

Il fallait reconnaître un des algorithmes étudiés en classe pour la conversion décimal $\rightarrow$ binaire.